sgl-kernel/csrc/cpu/gemm.cpp

```diff
@@ -474,6 +474,160 @@ at::Tensor convert_weight_packed(at::Tensor& weight) {
+              uint8_t src_1 = *(in_ptr + src_idx_1);
+              uint8_t src_2 = *(in_ptr + src_idx_2);
+              uint8_t dst = (src_1 & 0x0f) | ((src_2 & 0x0f) << 4);
+              *(out_ptr + dst_idx) = dst;
+            }
+          }
+        }
+      }
+      // compensation [block_n]
+      for (int nb = 0; nb < block_n; nb++) {
+        *(comp_out_prt + nb) = *(comp_in_prt + nb);
+      }
+    }
+  });
+
+  return std::make_tuple(std::move(blocked_weight), std::move(blocked_scales), std::move(blocked_qzeros));
+}
+
+std::tuple<at::Tensor, at::Tensor> autoawq_to_int4pack(
+    at::Tensor qweight,  // (*, K, N / 8), int32
+    at::Tensor qzeros)   // (*, K / group_size, N / 8), int32
+{
+  // bitshifts: [0, 4, 1, 5, 2, 6, 3, 7] * 4
+  auto bitshifts = at::tensor({0, 4, 1, 5, 2, 6, 3, 7}, at::kInt) * 4;
+  // qweight: assumed shape [..., K, N/8] (int32)
+  auto qweight_unsq = qweight.unsqueeze(-1);  // [..., K, N/8, 1]
+  auto shape = qweight_unsq.sizes().vec();    // shape: [A, B, C, 1]
+  shape[3] = 8;
+  auto unpacked = at::bitwise_right_shift(qweight_unsq, bitshifts) & 0xF;
+  auto qweight_final = unpacked.flatten(-2).transpose(-1, -2).to(at::kByte);
+
+  auto qzeros_unsq = qzeros.unsqueeze(-1);
+  auto qzeros_unpacked = at::bitwise_right_shift(qzeros_unsq, bitshifts) & 0xF;
+  auto qzeros_final = qzeros_unpacked.flatten(-2).to(at::kByte);
+
+  return std::make_tuple(qweight_final, qzeros_final);
+}
+
+std::tuple<at::Tensor, at::Tensor, at::Tensor> convert_weight_packed_scale_zp(
+    at::Tensor qweight,  // (*, K, N / 8), int32
+    at::Tensor qzeros,   // (*, K / group_size, N / 8), int32
+    at::Tensor scales    // (*, K / group_size, N), bfloat16
+) {
+  auto res = autoawq_to_int4pack(qweight, qzeros);
+  auto _qweight = std::get<0>(res);
+  auto _qzeros = std::get<1>(res);
+  auto _scales = scales;
+  _qzeros = _qzeros.transpose(-2, -1).contiguous();  // .T
+  _scales = _scales.transpose(-2, -1).contiguous();
+  if (_qweight.dim() == 3) {  // Dim=3 for MOE packing, TODO: refine a unified loop
+    int E = _qweight.size(0);
+    int K = _qweight.size(2);
+    int G = _scales.size(2);
+    int group_size = K / G;
+    int block_k = group_size > 128 ? 128 : group_size;
+    int block_n = block_size_n();
+    int Nc = _qweight.size(1) / block_n;
+    int Kc = K / block_k;
+    int64_t buffer_size_nbytes = block_k * block_n / 2 + block_n * sizeof(int32_t);
+    auto blocked_weight = at::empty({E, Nc, Kc, buffer_size_nbytes}, _qweight.options());
+    auto blocked_scales = at::empty({E, Nc, G, block_n}, _scales.options()).to(at::kFloat);
+    auto blocked_qzeros = at::empty({E, Nc, G, block_n}, _qzeros.options()).to(at::kChar);
+    for (int i = 0; i < _qweight.size(0); i++) {
+      auto res_ = convert_int4_weight_packed_with_compensation(_qweight[i], _scales[i], _qzeros[i]);
+      blocked_weight[i] = std::get<0>(res_);
+      blocked_scales[i] = std::get<1>(res_);
+      blocked_qzeros[i] = std::get<2>(res_);
```

Comment:
keep in mind that this type of operation is a out place copy in torch, it is going to be very slow.

we can keep as it is right now, as it won't affect benchmark performance, but model loading will be slow.